### ISSUE TITLE
Plugin crashing and stopped tracking play time

### ISSUE THREAD

**Issue by DarkNeooo** [tylerbrawl/Galaxy-Plugin-Rockstar#43]
After last Rockstar Launcher's update something went wrong and plugin stopped tracking time. I see information that I didn't play RDR2 for 3 days. And there's another issue - after updating plugin to 0.3.4 it started to launch games, but plugin is crashing every time when I hit _Play_ button. I have to click _Retry_ in Settings to restart plugin and I think only then the game starts.

**Comment by tylerbrawl** [comment]
Could you post your latest log file? I have yet to see this for myself.

**Comment by tylerbrawl** [comment]
I might actually have an explanation for this. Are you using a language other than English on your computer? I believe that I used a method for checking if the Rockstar Games Launcher is running that would break if the user did not have English as their language. 

I will post a potential fix below. If the attached build works, then I will close this issue and submit it as a release.

[plugin.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3859658/plugin.zip)

**Comment by DarkNeooo** [comment]
I'm using English as my system language, and unfortunately your fix doesn't work for me (I tried it anyway). 

Latest log for your plugin:
[plugin-rockstar-774732b5-69c4-405c-b6c9-92cd55740cfe.log](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3859751/plugin-rockstar-774732b5-69c4-405c-b6c9-92cd55740cfe.log)

If you need any of the other logs, tell me and I'll post it.

**Comment by tylerbrawl** [comment]
That is very interesting. Although there was an exception relating to importing the user's friends, this does not seem to be related to your issue. Unfortunately, the log file that you submitted does not show any activity relating to launching games. 

Would you be able to attempt to launch a game, wait for the plugin to crash, and then submit a zip file containing all of this plugin's log files? (If this log file was indeed submitted after you attempted to launch a game, then perhaps the problem _is_ related to importing friends.)

**Comment by DarkNeooo** [comment]
There is launch request at 2019-11-18 17:57:47,412, but maybe it isn't an activity you were talking about.

Here are two logs after I launched game another 2 times:
[logs.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3859909/logs.zip)

Ok... so I tested it a little bit more and now I see that plugin is launching game despite crashing. 
In addition, the games from Rockstar Launcher remain in my GOG Galaxy library, but _Play_ button is greyed out until I restart the plugin by clicking _Retry_.

**Comment by tylerbrawl** [comment]
I can see now that you did indeed make a request to launch a game, but the plugin crashed before it could finish. The strangest thing is that there is no exception listed in the log at all; it just cuts off as soon as the request to launch the game is announced. I have personally never experienced this myself, so it will be difficult attempting to find a fix. I am currently away from my computer right now, so it will be some time before I make another post, but I will see what I can do later.

In the meantime, perhaps you could look through the log files that Galaxy generates for itself (not those for the plugin specifically). Maybe something of use could be found in those?

**Comment by DarkNeooo** [comment]
Ok, I think you should check these two fragments. That's all I found in Galaxy's logs around the moment when plugin crashed and what seems to be related to chrashes. That fragment from _GalaxyClientService.log_ occured today only once, so it may not be related to it.

[log_fragments.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3860422/log_fragments.zip)

I hope this will help. And here are all the logs, you can try to find something, because you know what to look for.

[galaxylogs.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3860427/galaxylogs.zip)

**Comment by tylerbrawl** [comment]
Sadly, despite all of these log files, I still have no idea what is going on. I have included a build that might log what causes the crash in the Rockstar plugin log file. Would you be able to run it, wait for the plugin to crash after launching a game, and then send the latest Rockstar plugin log file(s) to me?

[plugin.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3860867/plugin.zip)

**Comment by DarkNeooo** [comment]
There is launch request at 2019-11-19 06:12:32,820, but unfortunately I don't see anything new.

[plugin-rockstar-774732b5-69c4-405c-b6c9-92cd55740cfe.log](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3862012/plugin-rockstar-774732b5-69c4-405c-b6c9-92cd55740cfe.log)

**Comment by tylerbrawl** [comment]
It seems like the plugin is failing to run the Windows ```tasklist``` command. Would you be able to enter the following into the Windows command prompt while playing a game (RDR 2 is used in this example):

```shell
tasklist /FI "IMAGENAME eq RDR2.exe " /FI "STATUS eq running" /FO LIST
```

If you are getting output displaying the image name, PID, etc., then the command is running correctly; if not, then perhaps your installation of the Windows ```tasklist.exe``` file is corrupted. If the command is indeed working correctly, then perhaps it needs to be sent directly through the shell.

I apologize for having you do all of this extra work, but this issue seems to be quite rare.

**Comment by DarkNeooo** [comment]
The command is working properly. 

Don't apologize - it's ok, I'm glad to help. It's strange that I'm the only one with the problem (or I'm the only one reporting it). The crashes started with version 0.3.4, before that everything was ok.

If you have any ideas and I can help you somehow then there's no problem, tell me what do you need.

**Comment by tylerbrawl** [comment]
If the command is working properly, then your Windows installation is fine. Maybe the ```tasklist``` command needs to be run directly through the shell for you? This typically is not required, but since the command works when run directly through the command prompt, it might fix it. Go ahead and test the attached build.

[plugin.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3872623/plugin.zip)

**Comment by DarkNeooo** [comment]
I tried the new build and unfortunately it didn't work :/

**Comment by tylerbrawl** [comment]
I may have found a solution. The documentation for the ```subprocess``` module states that referring to ```stdout``` directly from the process call can cause a deadlock in certain situations. (The page in particular is located at  https://docs.python.org/2/library/subprocess.html#subprocess.Popen.communicate.) This would mean that the following code from the original implementation could have caused your unlogged crashes:
```python
async def game_pid_from_tasklist(self, title_id) -> str:
        pid = None
        find_actual_pid = subprocess.Popen(
            f'tasklist /FI "IMAGENAME eq {games_cache[title_id]["launchEXE"]} " /FI "STATUS eq running" /FO LIST',
            stdout=subprocess.PIPE)

        for line in find_actual_pid.stdout:
            new_line = line.decode(locale.getpreferredencoding())
            if "PID" in new_line:
                pid = [str(s) for s in new_line.split() if s.isdigit()][0]
                break
        return pid
```
The troublesome part in particular is this:
```python
for line in find_actual_pid.stdout:
```
This actually makes sense when considering that the ```ROCKSTAR_LAUNCHER_PATCHER_PID``` message was never written to the log, even though it is supposed to be once ```game_pid_from_tasklist``` finished.

I edited this method in commit 45ef38a4391fe0cd64399bf914bd90f1ffa0090b to use ```Popen.communicate()``` instead of ```Popen.stdout```. Although the potential fix has not been pushed to the ```master``` branch, the build below incorporates it. Let's hope that it works this time.

[plugin.zip](https://github.com/tylerbrawl/Galaxy-Plugin-Rockstar/files/3876884/plugin.zip)

**Comment by DarkNeooo** [comment]
I wish I have some good news, but it doesn't work either. I'm starting to think that it has to be a problem on my side, but I don't know what it is. I've reinstalled GOG and your plugin. Meanwhile, there has been another update to Rockstar Launcher. And despite all of this the plugin is still crashing when I launch the game.

**Comment by tylerbrawl** [comment]
In that case, I honestly have no idea why the plugin is not working. Could it be a potential conflict between plugins?

**Comment by DarkNeooo** [comment]
I don't think so. Your plugin is the only one that I have downloaded on my own. And before that update to Rockstar Launcher that broke the plugin (0.3.3) everything was ok. It's weird.

**Comment by tylerbrawl** [comment]
Thank you for doing that. The documents for the ```subprocess``` module state that you should rarely if ever set the ```shell``` parameter to ```True```, and leaving it as ```False``` should be sufficient for a task like this. I also initially experienced issues with launching a game directly from the shell. It turns out that you need quotation marks around the game path, since there are likely spaces contained in the path of your game. The simplest fix would be to change this:
```python
subprocess.call(f'{game_path} -launchTitleInFolder "{path}" @commandline.txt', stdout=self.FNULL, stderr=self.FNULL, shell=False)
```
with this:
```python
subprocess.call(f'"{game_path}" -launchTitleInFolder "{path}" @commandline.txt', stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
```

This does not solve the underlying problem of the ```subprocess``` module crashing, but it should work as a workaround for you.

**Comment by DarkNeooo** [comment]
When I changed 'shell' to 'True' and added quotation marks to '{game_path}' then the plugin launched the game, but it still crashed.

**Comment by MoshiMoshi0** [comment]
As per docs `subprocess.call` waits for the process to finish so it blocks the script until you exit the game. Also the args for call/Popen should be a list of arguments not a string.

This should fix the issue:
```python
subprocess.Popen([game_path, '-launchTitleInFolder', path, '@commandline.txt'], stdout=self.FNULL,
                        stderr=self.FNULL, shell=False)
```

I'm not sure if you need to use devnull for stdout/stderr, the default NONE might be enough.

**Comment by DarkNeooo** [comment]
It worked! The game started and the plugin didn't crash.

**Comment by tylerbrawl** [comment]
@MoshiMoshi0 I didn't know that about ```subprocess.call```; good catch. I guess the strangest thing is how it worked fine on my computer, but not for @DarkNeooo. Also, ```None``` can apparently be used in this case for ```stdout``` and ```stderr```, since the Rockstar Games Launcher does not actually output anything to the shell; thus, blocking would not be an issue. Still, I like being more explicit about things like that, personally.

Lastly, using a sequence instead of a string is a good idea, on the off chance that Rockstar somehow ports their games and launcher to Linux/Mac. (This will probably never happen, but we can always hope.) Thanks again for your help!
